### PR TITLE
[WIP] Structured items

### PR DESCRIPTION
**Pull request by kvark**
The PR contains core missing pieces of #5: rich instructions and top-level items, such as `Module`.

### CODE REVIEW

**Comment by kvark** [comment]
I think the biggest thing missing is basically a `function(Context, DataInstruction) ->StructuredInstruction`. It could be implemented as a few auto-generated methods:
```rust
impl Context {
  pub fn lift_type(&mut self, raw: &Instruction) -> TypeToken;
  pub fn lift_extension(&mut self, raw: &Instruction) -> Extension;
  pub fn lift_operand(&mut self, <something>) -> Operand;
  ..etc
}
```

**Comment by kvark** [comment]
Looking at this some more... I'm able to generate the lift() functions that process and match operands, aka:
```rust
    pub fn lift_memory_model(
        &mut self,
        raw: &mr::Instruction,
    ) -> Result<structs::MemoryModel, LiftError> {
        if raw.class.opcode as u32 != 14u32 {
            return Err(LiftError::OpCode);
        }
        let mut operands = raw.operands.iter();
        Ok(structs::MemoryModel {
            addressing_model: (match operands.next() {
                Some(&mr::Operand::AddressingModel(ref value)) => Some(value.clone()),
                Some(_) => Err(OperandError::Wrong)?,
                None => None,
            })
            .ok_or(OperandError::Missing)?,
            memory_model: (match operands.next() {
                Some(&mr::Operand::MemoryModel(ref value)) => Some(value.clone()),
                Some(_) => Err(OperandError::Wrong)?,
                None => None,
            })
            .ok_or(OperandError::Missing)?,
        })
    }
```

My problem is that we need to enforce the semantics that is in the spec but nowhere to be seen in the schema, e.g.:
> Entry Point must be the Result <id> of an OpFunction instruction.
> Interface is a list of <id> of global OpVariable instructions

In the schema they are just `IdRef` things:
```json
        { "kind" : "IdRef",                              "name" : "'Entry Point'" },
        { "kind" : "LiteralString",                      "name" : "'Name'" },
        { "kind" : "IdRef",          "quantifier" : "*", "name" : "'Interface'" }
```

So there appears to be incomplete information in the schema to build the proper structured API. We have 2 ways to proceed from here:
  1. just implement all/most instructions manually, use the exact typing corresponding to the spec
  2. associate operand names with types in the code generator

@antiagainst do you have a feel on which approach would work best for us? ^

**Comment by kvark** [comment]
@antiagainst I'm going with approach (2) currently, i.e.:
```rust
        match operand.name.trim_matches('\'') {
            "Length" => quote! { ConstantToken },
            "Entry Point" => quote! { FunctionToken },
            "Interface" => quote! { VariableToken },
            _ => quote! { TypeToken },
        }
```
So far, it seems to me that naming is fairly consistent and repetitive, so adding such a match is still a big win over manually writing the code.

Another issue I'm facing is that content in SPIRV appears to not be sequential. An OpEntryPoint point to OpFunction, for example. This is annoying because we want to be populating our context graph with function/instruction/constant/etc tokens, and we can use one unless we processed it first. I'll try to see if we can process all the functions first, before going for the mode setting ops, and whether this is big issue within a function/block as well.

**Comment by antiagainst** [comment]
Yes the JSON grammar was not a full-fledged encoding of the syntax validation rules. (My bad initially defining it that way; it was more meaning to be supporting a data driven approach in SPIRV-Tools.) And you pointed out the most noticeable one: IdRef. I initially thought about adding more sub cases to it but never got to it.

Generally I'm in favor of more autogen. (You can know that which all the codegen stuff I've in this project; although codegen means another thing in compiler so we should probably rename it. ;-P) So matching names as a start sounds good to me. I think a more principled solution would be to enhance the SPIR-V JSON grammar to encode this info. (I'm actually needing other stuff that is not in the upstream; category is added by me for this repo specifically and so you see we are not using an upstream grammar here, which is bad because need to manually merge with each version update.)

**Comment by antiagainst** [comment]
There are a few places that SPIR-V has forward references. The [sections](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_logicallayout_a_logical_layout_of_a_module) before types and global values are all "metadata" instructions; meaning they attach additional information to some other instruction. So frequently they do forward referencing, esp. entry point, execution mode, debugging instruction, annotations. (We can also have `OpTypeForwardPointer` forward references a type...) Sounds good to me to parse the "main" instructions first (the types, global values, and functions) and then getting back to those instructions before. (We have the full MR (which should be named into as DR) representation here so it should be doable.) Also we should fold those metadata instructions into their main instruction if possible. That's why I have a decoration field for most of the instructions.

**Comment by kvark** [comment]
This is obviously incomplete, WIP, etc, but at least it's at the point where it compiles. I don't think having more stuff piled up on top would be a good way to merge later. @antiagainst - do you think you could review this and consider landing if it's done in the preferred way?

File: codegen/sr.rs [hunk 1]
```diff
@@ -21,25 +21,19 @@ use proc_macro2::{Ident, Span, TokenStream};
...
-    } else if kind == "PairLiteralIntegerIdRef" {
-        quote! { (u32, spirv::Word) }
-    } else if kind == "PairIdRefLiteralInteger" {
-        quote! { (spirv::Word, u32) }
-    } else if kind == "PairIdRefIdRef" {
-        quote! { (spirv::Word, spirv::Word) }
-    } else {
-        let kind = Ident::new(kind, Span::call_site());
-        quote! { spirv::#kind }
+    match kind {
```
Comment (antiagainst) [inline codegen/sr.rs]: This is just refining existing implementation and not related to the new feature we are implementing here, right? In the future, it would be nice to have separate PRs.

File: codegen/sr.rs [hunk 2]
```diff
@@ -280,51 +271,192 @@ pub fn gen_sr_type_creation(grammar: &structs::Grammar) -> String {
...
+                    vec.push(value);
+                }
+                vec
+            }
+        },
+        other => panic!("wrong quantifier: {}", other),
+    }
+}
+
+pub fn gen_sr_structs_and_lifts(grammar: &structs::Grammar) -> (String, String) {
```
Comment (antiagainst) [inline codegen/sr.rs]: Can we have some comments regarding this function?

(I wasn't doing good on this previously given it's just me playing with this repo. Now it's intended to be developed by more people so some documentation could really save newcomers efforts in understanding existing codebase. :)

File: rspirv/Cargo.toml [hunk 3]
```diff
@@ -20,6 +20,7 @@ travis-ci = { repository = "gfx-rs/rspirv" }
 num = "0.2"
 derive_more = "0.7"
 clippy = { version = "0.0", optional = true }
+fxhash = "0.2"
```
Comment (antiagainst) [inline rspirv/Cargo.toml]: Nit: can you sort these dependencies alphabetically?

File: codegen/sr.rs [hunk 4]
```diff
@@ -260,14 +254,11 @@ pub fn gen_sr_type_creation(grammar: &structs::Grammar) -> String {
...
             quote! {
-                pub fn #func_name #param_list -> TypeToken {
-                    let t = Type { ty: TypeEnum::#symbol #init_list, decorations: Vec::new() };
-                    if let Some(index) = self.types.iter().position(|x| *x == t) {
-                        TypeToken::new(index)
-                    } else {
-                        self.types.push(t);
-                        TypeToken::new(self.types.len() - 1)
-                    }
+                pub fn #func_name #param_list -> Token<Type> {
```
Comment (antiagainst) [inline codegen/sr.rs]: Here too. It would be nice to have a separate CL for this change and the autogen touched files.

File: rspirv/sr/context.rs [hunk 5]
```diff
@@ -12,11 +12,94 @@
...
+    items::{Function, Variable},
+    structs,
+    LiftError, OperandError,
+};
+use crate::mr;
+
+type FastHashMap<K, V> = HashMap<K, V, BuildHasherDefault<FxHasher>>;
+type Index = u32;
+
+/// A structure holding some kind of a structured SPIRV data that can be
```
Comment (antiagainst) [inline rspirv/sr/context.rs]: Nit: SPIR-V is the formal spelling. Unless we are in code where we cannot use -, let's use the formal name when possible.

Comment (antiagainst) [reply rspirv/sr/context.rs]: ... some kind of SPIR-V constructs that can ...

File: rspirv/sr/context.rs [hunk 6]
```diff
@@ -29,24 +112,55 @@ use sr::types::{StructMember, TypeEnum};
...
+        &self.types.map[&token.index]
+    }
+
+    /// Returns the reference to the real constant represented by the given token.
+    pub fn get_constant(&self, token: Token<Constant>) -> &Constant {
+        // Note: we assume the vector doesn't shrink so we always have a valid index.
+        &self.constants.map[&token.index]
+    }
+
+    pub fn get_function(&self, token: Token<Function>) -> &Function {
```
Comment (antiagainst) [inline rspirv/sr/context.rs]: Can we have documentations for these public methods?

File: rspirv/sr/instructions.rs [hunk 7]
```diff
@@ -0,0 +1,7 @@
+use spirv;
+use super::{
+	context::Token,
```
Comment (antiagainst) [inline rspirv/sr/instructions.rs]: Format

File: rspirv/sr/items.rs [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+impl From<LiftError> for ConvertionError {
+    fn from(error: LiftError) -> Self {
+        ConvertionError::Lift(error)
+    }
+}
+
+impl Module {
+    pub fn from_data(module: &mr::Module) -> Result<Self, ConvertionError> {
```
Comment (antiagainst) [inline rspirv/sr/items.rs]: Nit: from_data_representation

File: rspirv/sr/mod.rs [hunk 9]
```diff
@@ -14,12 +14,38 @@
...
 mod context;
-mod autogen_decoration;
+mod instructions;
+mod items;
+pub mod structs;
 mod types;
+
+
+#[derive(Clone, Debug)]
+pub enum OperandError {
```
Comment (antiagainst) [inline rspirv/sr/mod.rs]: Can we have documentation for public APIs?

File: rspirv/sr/structs.rs [hunk 10]
```diff
@@ -0,0 +1,7 @@
+use spirv;
```
Comment (antiagainst) [inline rspirv/sr/structs.rs]: `structs` is not a very descriptive name for this module. It's too general and it implies data representation (which is another representation). Can we have a better one? What about `op`?

File: rspirv/sr/autogen_instruction.rs [hunk 11]
```diff
@@ -25,656 +68,627 @@ enum Instruction {
...
     },
-    Extension {
-        name: String,
-    },
-    ExtInstImport {
-        name: String,
-    },
     ExtInst {
-        set: spirv::Word,
+        set: Token<Type>,
```
Comment (antiagainst) [inline rspirv/sr/autogen_instruction.rs]: Shouldn't all these be `Token<Instruction>`?

**Comment by kvark** [comment]
@antiagainst This PR becomes increasingly difficult to advance without either of #49 or #55 to be merged.

**Comment by antiagainst** [comment]
Just wanted to double check here. So this PR is not expected to be reviewed but expected to be split into smaller ones right?

**Comment by kvark** [comment]
@antiagainst correct. It's technically in the "Draft" status, which isn't ready for review.
I think I'm getting close to catching up to this PR...

**Comment by antiagainst** [comment]
Cool, thanks!

**Comment by Jasper-Bekkers** [comment]
This PR has been stale for about a year, mind if we close it?
